src/storage/cache_proxy/cluster_manager.rs

```diff
@@ -0,0 +1,1096 @@
+//! The utilities of distribute cache cluster management
+
+use std::sync::Arc;
+use std::time::Duration;
+
+use anyhow::Context;
+use arc_swap::{ArcSwap, ArcSwapOption};
+use futures::StreamExt;
+use tracing::{debug, warn};
+
+use crate::async_fuse::memfs::kv_engine::etcd_impl::Session;
+use crate::async_fuse::memfs::kv_engine::{KVEngine, SetOption};
+use crate::async_fuse::memfs::kv_engine::{KVEngineType, KeyType, ValueType};
+use crate::common::error::{DatenLordError, DatenLordResult};
+
+use super::node::{MasterNodeInfo, Node, NodeStatus};
+use super::ring::Ring;
+
+/// The timeout for current node's session
+const SESSION_TIMEOUT_SEC: u64 = 10;
+
+/// This struct is used to interact with etcd server and manager the cluster.
+#[derive(Debug, Clone)]
+#[allow(dead_code)]
+pub struct ClusterManager {
+    /// inner struct
+    inner: Arc<ClusterManagerInner>,
+}
+
+#[allow(dead_code)]
+impl ClusterManager {
+    /// Create new cluster manager
+    pub fn new(kv_engine: Arc<KVEngineType>, node: Node) -> Self {
+        let inner = Arc::new(ClusterManagerInner::new(kv_engine, node));
+        Self { inner }
+    }
+
+    /// Get current node
+    #[must_use]
+    pub fn get_node(&self) -> Node {
+        self.inner.get_node()
+    }
+
+    /// Get current hashring
+    pub async fn get_ring(&self) -> DatenLordResult<Ring<Node>> {
+        self.inner.get_ring(false).await
+    }
+
+    /// Get all nodes
+    pub async fn get_nodes(&self) -> DatenLordResult<Vec<Node>> {
+        self.inner.get_nodes(false).await
+    }
+
+    ///  Stop the cluster manager
+    pub fn stop(&self) {
+        // In this step, we just to clean session and force down the cluster manager
+        self.inner.clean_sessions();
```

Comment:
How to foce shutdown?  after clean_session, run() will continue in the loop.

Reply:
The shutdown here is for test cases, if I need to force a shutdown I tend to set a flag and wait until the clean session is done loop will check the status of this flag.

Reply:
Stop `watch_xxx` task by channel event.